semantics/cpp14/language/translation/decl/enum.k

```diff
@@ -80,23 +80,52 @@ module CPP-DECL-ENUM
+     rule declareEnumerators(_, _, _, Next::Int, PrevType::CPPSimpleType,
+               ListItem(Enumerator(_, NoExpression() => prv(Next, noTrace, type(PrevType)))) _::List, false, false)
+          requires inRange(Next, type(PrevType))
+
+     // Second or more enumerator, underlype type not fixed, value does not fit to previous type but fits to some.
+     rule (.K => UNSPEC("TDE3", "Value does not fit to type of preceding enumerator")) ~> (
+          declareEnumerators(E::CPPEnumType, Min::Int, Max::Int, Next::Int, PrevType::CPPSimpleType, ListItem(Enumerator(X::CId, NoExpression())) L::List, false, false)
+          => recoverFromError(
+          declareEnumerators(E, Min, Max, Next::Int, PrevType, ListItem(Enumerator(X, prv(Next, noTrace, type(firstSuitableType(Next, Next, allIntegralTypes))))) L, false, false)
+          ))
+          requires notBool inRange(Next, type(PrevType)) andBool firstSuitableType(Next, Next, allIntegralTypes) =/=K no-type
+
+     rule (.K => ILL("TDE4", "Incremented enumerator value is not representable in any integer type")) ~>
+          declareEnumerators(_, _, _, Next::Int, PrevType::CPPSimpleType, ListItem(Enumerator(_, NoExpression())) _::List, false, false)
+          requires notBool inRange(Next, type(PrevType)) andBool firstSuitableType(Next, Next, allIntegralTypes) ==K no-type
+
+     rule declareEnumerators(_, _, _, _, _, ListItem(Enumerator(_, prv(_, _, T:CPPEnumType => underlyingType(T)))) _::List, _, _)
      context declareEnumerators(_, _, _, _, _, ListItem(Enumerator(_, HOLE:Expr)) _::List, _, _)
 
      // Type of an enumerator inside the declaration is enum's underlying type
      // if it is fixed, and type of initializing expression otherwise,
      // Hence we need to store the type of underlying expression...
-     rule declareEnumerators(T::CPPEnumType, Min::Int, Max::Int, Next::Int, _, ListItem(Enumerator(X::CId, prv(V::Int, _, t(_, _, EType::CPPSimpleType)) #as Pr::PRVal)) L::List, Fixed::Bool, true)
-          => declareEnumerator(getEnumId(T), X, Pr, Fixed, underlyingType(T))
-          ~> declareEnumerators(T, V, V, V +Int 1, chooseEnumeratorType(EType, V +Int 1), L, Fixed, false)
-
-     rule declareEnumerators(T::CPPEnumType, Min::Int, Max::Int, Next::Int, _, ListItem(Enumerator(X::CId, prv(V::Int, _, t(_, _, EType::CPPSimpleType)) #as Pr::PRVal)) L::List, Fixed::Bool, false)
+     rule declareEnumerators(T::CPPEnumType, Min::Int, Max::Int, Next::Int, _, ListItem(Enumerator(X::CId, prv(V::Int, _, t(_, _, EType::CPPSimpleType)) #as Pr::PRVal)) L::List, Fixed::Bool, First::Bool)
           => declareEnumerator(getEnumId(T), X, Pr, Fixed, underlyingType(T))
-          ~> declareEnumerators(T, minInt(Min, V), maxInt(Max, V), V +Int 1, chooseEnumeratorType(EType, V +Int 1), L, Fixed, false)
-
+          ~> declareEnumerators(T, ite(First, V, minInt(Min, V)), ite(First, V, maxInt(Max, V)), V +Int 1, EType, L, Fixed, false)
```

Comment:
ah yes. This is a known issue in the ocaml backend. When using a boolean variable in boolean operators (ie and, or, not, and if/then/else), please explicitly type the boolean variable (i.e. `First:Bool`).